doc/cprover-manual/goto-harness.md

```diff
@@ -0,0 +1,312 @@
+
+int function_to_test(int some_argument)
+{
+  assert(some_argument == 0);
+  return some_argument;
+}
+```
+
+We first need to generate a GOTO binary.
+
+```sh
+$ goto-cc -o main.gb main.c
+```
+
+Then we can call `goto-harness` on the generated GOTO binary
+to get a modified one that contains the harness function:
+
+```sh
+$ goto-harness \
+--harness-function-name harness \
+--harness-type call-function \
+--function function_to_test \
+main.gb \
+main-with-harness.gb
+```
+
+The options we pass to `goto-harness` are:
+
+* `harness-function-name` is the name of the function generated by the harness
+  generator (this needs to be specified for all harness generators).
+* `harness-type` is the type of harness to generate (`call-function` is the
+  function-call harness generator)
+* `function` is the name of the function that gets instrumented
+* then we pass the input GOTO-binary and a name for the output GOTO binary.
+
+What comes of this is a GOTO binary roughly corresponding to the following C
+pseudocode:
+
+```C
+// modified_goto_binary.c
+
+#include <assert.h>
+
+int function_to_test(int some_argument)
+{
+  assert(some_argument == 0);
+  return some_argument;
+}
+
+void harness(void)
+{
+  int argument = nondet_int();
+  function_to_test(argument);
+}
+```
+
+You can pass that modified GOTO binary to `cbmc` and select `harness` as the
+entry function for analysis, or further modify it with `goto-harness` or
+`goto-instrument` or other tools.
+
+The example above demonstrates the most simple case, which is roughly the same
+as the entry point `cbmc` automically generates for functions. However, the
+`function-call` harness can also non-deterministically initialise global
+variables, array and struct elements, consider this more complicated example:
+
+```C
+// list_example.c
+#include <assert.h>
+#include <stdlib.h>
+
+typedef struct linked_list linked_list;
+struct linked_list {
+    int value;
+    linked_list *next;
+};
+
+linked_list *global_linked_list;
+
+/// initialize all values in the global
+/// list to 0
+void initialize_global(void)
+{
+  for(linked_list *ll = global_linked_list;
+    ll != NULL;
+    ll = ll->next)
+  {
+    ll->value = 0;
+  }
+}
+
+/// try to initialize all values in the linked list
+/// to 0 - but this version contains two bugs,
+/// as it won't work with nullpointer arguments
+/// and it will also not initialize the last element
+void initialize_other(linked_list *ll)
+{
+    do {
+        ll->value = 0;
+        ll = ll->next;
+    } while(ll->next != NULL);
+}
+
+void check_list(linked_list *list_parameter)
+{
+    assert(list_parameter != global_linked_list);
+    initialize_global();
+    initialize_other(list_parameter);
+    linked_list *global_cursor = global_linked_list;
+    linked_list *parameter_cursor = list_parameter;
+
+    // global list should be a prefix of the parameter now,
+    // or the other way round
+    while(global_cursor != NULL && parameter_cursor != NULL)
```

Comment:
// global list should be a prefix of the parameter now,
    // or the other way round

Not sure what this means.

Reply:
the above methods ensure that both linked lists are some linked lists that are filled with `0` values, so one is definitely the prefix of the other, but we don't know which (of course they might both be prefixes of the other if they're equal). This is documenting what is expected here. As you rightly pointed out there is a deliberate bug in the implementation that causes one of the lists to be incorrectly initialised so this isn't actually _true_, but it's what we're asserting in the loop.

Reply:
So the bug would not appear under all circumstances. And this is intentional. Okay, thanks.